File: Source/AlphaTab/Rendering/Glyphs/TabBeatPreNotesGlyph.cs
using AlphaTab.Model;

namespace AlphaTab.Rendering.Glyphs
{
    internal class TabBeatPreNotesGlyph : BeatGlyphBase
    {
        public override void DoLayout()
        {
            if (Container.Beat.BrushType != BrushType.None && !Container.Beat.IsRest)
            {
                AddGlyph(new TabBrushGlyph(Container.Beat));
                AddGlyph(new SpacingGlyph(0, 0, 4 * Scale));
            }

            base.DoLayout();
        }
    }
}

File: packages/alphatab/src/rendering/glyphs/TabBeatPreNotesGlyph.ts
import { BeatSubElement } from '@coderline/alphatab/model/Beat';
import { BrushType } from '@coderline/alphatab/model/BrushType';
import { BeatGlyphBase } from '@coderline/alphatab/rendering/glyphs/BeatGlyphBase';
import { SpacingGlyph } from '@coderline/alphatab/rendering/glyphs/SpacingGlyph';
import { TabBrushGlyph } from '@coderline/alphatab/rendering/glyphs/TabBrushGlyph';

/**
 * @internal
 */
export class TabBeatPreNotesGlyph extends BeatGlyphBase {
    public override doLayout(): void {
        if (this.container.beat.brushType !== BrushType.None && !this.container.beat.isRest) {
            this.addEffect(new TabBrushGlyph(this.container.beat));
            this.addNormal(new SpacingGlyph(0, 0, this.renderer.smuflMetrics.preNoteEffectPadding));
        }
        super.doLayout();
    }

    protected override get effectElement() {
        return BeatSubElement.GuitarTabEffects;
    }
}
